packages/safe-ds-vscode/src/extension/eda/apis/runnerApi.ts

```diff
@@ -0,0 +1,415 @@
+import { Base64Image, Column, Profiling, ProfilingDetailStatistical, Table } from '@safe-ds/eda/types/state.js';
+import { SafeDsServices, messages } from '@safe-ds/lang';
+import { LangiumDocument, AstNode } from 'langium';
+import { printOutputMessage } from '../../output.ts';
+import * as vscode from 'vscode';
+import crypto from 'crypto';
+import { getPipelineDocument } from '../../mainClient.ts';
+import { CODEGEN_PREFIX } from '../../../../../safe-ds-lang/src/language/generation/safe-ds-python-generator.ts';
+
+export class RunnerApi {
+    services: SafeDsServices;
+    pipelinePath: vscode.Uri;
+    pipelineName: string;
+    baseDocument: LangiumDocument<AstNode> | undefined;
+    placeholderCounter = 0;
+
+    constructor(services: SafeDsServices, pipelinePath: vscode.Uri, pipelineName: string) {
+        this.services = services;
+        this.pipelinePath = pipelinePath;
+        this.pipelineName = pipelineName;
+        getPipelineDocument(this.pipelinePath).then((doc) => {
+            // Get here to avoid issues because of chanigng file
+            // Make sure to create new instance of RunnerApi if pipeline execution of fresh pipeline is needed
+            // (e.g. launching of extension on table with existing state but no current panel)
+            this.baseDocument = doc;
+        });
+    }
+
+    private async addToAndExecutePipeline(pipelineExecutionId: string, addedLines: string): Promise<void> {
+        return new Promise(async (resolve, reject) => {
+            if (!this.baseDocument) {
+                reject('Document not found');
+                return;
+            }
+
+            const documentText = this.baseDocument.textDocument.getText();
+
+            // Find pattern "pipeline <pipelineName> {" and add the SDS code before the closing bracket of it
+            const pipelineStart = documentText.indexOf('pipeline ' + this.pipelineName);
```

Comment:
That's not going to work if there is extra whitespace or comments or any other hidden tokens between the `pipeline` keyword and the name. Generally, you should avoid parsing the document text again, that's what we have the parser created by Langium for. 

Instead, start with the AST node for the pipeline and get it's `$cstNdde`. From there, you can access the end position or offset of the pipeline, which is where you can add your code.